### PR TITLE
Add graphs and summary information for Illumina runs

### PR DESCRIPTION
**Pull request by apmasell**
JIRA: GLT-1336

### CODE REVIEW

**Comment by apmasell** [comment]
For testing purposes, build `runscanner-illumina` per the instructions and then do:
```
mkdir ~/mnt
sshfs -o idmap=user -o BatchMode=no hn.hpc.oicr.on.ca:/.mounts ~/mnt
 ./runscanner-illumina ~/mnt/labs/prod/archive/h1179/160406_SN7001179_0025_AC8CVBACXX/ | jq -r '("UPDATE Run SET metrics = '\''\(.metrics)'\'' WHERE runId = 1;")' | mysql
```
Add whatever parameters you need to `mysql`. You can then view run 1 and the graphs will be provided.

**Comment by callunity** [comment]
Following the build instructions causes this error on my Ubuntu machine: `Unable to locate package build-essentials`

**Comment by apmasell** [comment]
Updated.

**Comment by callunity** [comment]
In the first graph, what is meant by `Cycles` ? In the graphs coming from the Notification Server, it was a representation of `completedCycles / totalCycles` but now it's only a single number. Is there another way of seeing how many cycles have completed and how many remain?

It seems like there's a lot of white space to the right of the graphs, which is making them look scrunched. Is there a way to reduce this?

**Comment by apmasell** [comment]
I think the cycles can be changed. There's a cycle state object that has extracted, called, Q-scored. That number is Q-scored. There's also a total cycles. Should I display all 4? I have no idea which one was used before.

I don't see the whitespace problem, but the layout of the graphs becomes fixed because the highcharts don't seem to be able to resize dynamically. I'm not sure what to do about this.

**Comment by callunity** [comment]
I was playing around with the CSS and wasn't able to come up with anything good for resizing the Highcharts either.

In staging, the run you chose displays `Cycles: 209/209`. Are there any numbers in the cycle state object that matches up with that? Does total cycles match up with that?

**Comment by apmasell** [comment]
They're all the same when the run is complete.

**Comment by apmasell** [comment]
I've added all the cycle states but they seem to be zero. I don't know if the cycle states information is missing from our runs. :-/

**Comment by apmasell** [comment]
The cycle stats are defined as a range, but it seems that start == end, so I'm just using end.

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java [hunk 1]
```diff
@@ -605,6 +609,57 @@ public boolean processNotification(Run source, int laneCount, String containerSe
...
+      target.setMetrics(source.getMetrics());
+      return true;
+    }
+
+    ObjectMapper mapper = new ObjectMapper();
+    ArrayNode sourceMetrics;
+    try {
+      sourceMetrics = mapper.readValue(source.getMetrics(), ArrayNode.class);
+    } catch (IOException e) {
+      log.debug("Impossible junk metrics were passed in for run " + target.getId(), e);
```
Comment (djcooke) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java]: Should probably be concerning. IOException may be thrown for other reasons too. Log at error level

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java [hunk 2]
```diff
@@ -605,6 +609,57 @@ public boolean processNotification(Run source, int laneCount, String containerSe
...
+      return false;
+    }
+    ArrayNode targetMetrics;
+    try {
+      targetMetrics = mapper.readValue(target.getMetrics(), ArrayNode.class);
+    } catch (IOException e) {
+      log.debug("The database is full of garbage metrics for run " + target.getId(), e);
```
Comment (djcooke) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java]: More concerning

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java [hunk 3]
```diff
@@ -605,6 +609,57 @@ public boolean processNotification(Run source, int laneCount, String containerSe
     return isNew;
   }
 
+  private boolean updateMetricsFromNotification(Run source, Run target) {
+    if (source.getMetrics() != null && source.getMetrics().equals(target.getMetrics())) return false;
+    if (source.getMetrics() == null && target.getMetrics() != null) {
+      return false;
```
Comment (djcooke) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java]: Can probably return false if source metrics are null, regardless of target metrics?

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java [hunk 4]
```diff
@@ -605,6 +609,57 @@ public boolean processNotification(Run source, int laneCount, String containerSe
...
+    targetMetricsMap.putAll(sourceMetricsMap);
+    ArrayNode combinedMetrics = mapper.createArrayNode();
+    combinedMetrics.addAll(targetMetricsMap.values());
+    try {
+      target.setMetrics(mapper.writeValueAsString(combinedMetrics));
+    } catch (JsonProcessingException e) {
+      log.error("Failed to save data just unserialised.", e);
+      return false;
+    }
+    return false;
```
Comment (djcooke) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java]: Seems like true should be returned at this point

**Comment by apmasell** [comment]
Fixed.

**Pull request merged**